theano/gpuarray/GpuMaxAndArgmax.py

```diff
@@ -0,0 +1,209 @@
+            all_axes.sort()
+            if all_axes == range(X.type.ndim):
+                axis = None
+        else:
+            all_axes = range(X.ndim)
+            axis_set = set(all_axes)
+        if axis is None:
+            axis = NoneConst.clone()
+        else:
+            axis = theano.tensor.as_tensor_variable(all_axes)
+            # assert axis.ndim == 1
+        inputs = [as_gpuarray_variable(X, context_name), axis]
+        # We keep the original broadcastable flags for dimensions on which
+        # we do not perform the max / argmax.
+        broadcastable = [b for i, b in enumerate(X.type.broadcastable)
+                         if i not in axis_set]
+        outputs = [GpuArrayType(X.type.dtype, broadcastable, context_name=context_name, name='max')(),
+                   GpuArrayType(self.argmax_dtype, broadcastable, context_name=context_name, name='argmax')()]
+        return Apply(self, inputs, outputs)
+
+    def perform(self, node, inputs, outputs):
+        X, axes = inputs
+        max, max_idx = outputs
+        if axes is None:
+            axes = tuple(range(X.ndim))
+        else:
+            axes = tuple(axes)
+            # axes = tuple(int(ax) for ax in axes)
+        max[0] = theano._asarray(numpy.max(X, axes), dtype=node.outputs[0].dtype)
+        # Numpy does not support multiple axes for argmax
+        # Work around
+        keep_axes = numpy.array([i for i in range(X.ndim) if i not in axes], dtype='int64')
+        # Not-reduced axes in front
+        transposed_x = numpy.transpose(X, numpy.concatenate((keep_axes, axes)))
+        kept_shape = transposed_x.shape[:len(keep_axes)]
+        reduced_shape = transposed_x.shape[len(keep_axes):]
+        new_shape = kept_shape + (numpy.prod(reduced_shape),)
+        reshaped_x = transposed_x.reshape(new_shape)
+
+        max_idx[0] = theano._asarray(numpy.argmax(reshaped_x, axis=-1), dtype='int64')
+
+    def c_headers(self):
+        return ['<numpy_compat.h>', '<gpuarray_helper.h>']
+
+    def c_header_dirs(self):
+        return [pygpu.get_include(), os.path.dirname(__file__)]
+
+    def c_code(self, node, name, input_names, output_names, sub):
+        # Recall: X, axes = input_names
+        # Recall: max, argmax = output_names
+        # Recall: fail = sub['fail']
+        max_typecode = pygpu.gpuarray.dtype_to_typecode(node.inputs[0].dtype)
+        argmax_typecode = pygpu.gpuarray.dtype_to_typecode(self.argmax_dtype)
+        # axes_ctype = pygpu.gpuarray.dtype_to_ctype(node.inputs[1].dtype)
+        axes_ctype = 'int64_t'
+        ret = """
+        GpuArray* %(name)s_input = &%(X)s->ga;
+        size_t %(name)s_input_ndim = PyGpuArray_NDIM(%(X)s);
+        """
+        if NoneConst.equals(node.inputs[1]):
+            ret += """
+            unsigned  %(name)s_redux_len = %(name)s_input_ndim;
+            unsigned* %(name)s_axes_to_reduce = new unsigned[%(name)s_redux_len];
+            for(unsigned i = 0; i < %(name)s_redux_len; ++i) {
```

Comment:
Our C style includes placing a space before the parenthesis for non-functions (`for`, `if`, ...)